### PR TITLE
Reintroduce `onReanimatedPropsChange`

### PR DESCRIPTION
**Pull request by Latropos**
## Summary

Reintroduce support for jsProps. 
Closes https://github.com/software-mansion/react-native-reanimated/issues/4317 and closes https://github.com/software-mansion/react-native-reanimated/issues/4674

It turns out that jsProps are still in use by many libraries like `react-native-lottie` and `react-native-svg`.

> **Note**
> This fix does not work for fabric 😢 However Lottie animations are not implemented for Fabric anyway 😎

<table>
<tr><td>BEFORE</td><td>AFTER</td></tr>

<tr><td>The white dot does not move ❌</td><td>The white dot moves ✅</td></tr>
<tr><td>

https://github.com/software-mansion/react-native-reanimated/assets/56199675/3c961690-383c-45dc-83ec-5801e13473bf

</td><td>

https://github.com/software-mansion/react-native-reanimated/assets/56199675/bf900564-cb5a-43cb-b2ef-2d21794727a1

</td></tr>
</table>

## Test plan

Tested on this example: https://github.com/acollazomayer/ReanimatedSVGBug

### CODE REVIEW

File: src/ReanimatedModule.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  },
+  async createNode(
+    _nodeID: number,
+    _config: Record<string, unknown>
+  ): Promise<void> {
+    // noop
+  },
+  async dropNode(_nodeID: number): Promise<void> {
+    // noop
+  },
```
Comment (tomekzaw) [inline src/ReanimatedModule.ts:30]: Why do we need this method?

**Comment by AlexanderEggers** [comment]
@Latropos Quick FYI, lottie-react-native v6 with fabric support was released a short while ago.

File: src/ReanimatedModule.ts (hunk 1 above)
Comment (Latropos) [reply src/ReanimatedModule.ts]: You're right, I'll keep only the ones that actually are in use

**Review by piaskowyk — CHANGES_REQUESTED** [review]
let's refactor the event registering logic into a separate file, just like we discussed offline.

**Comment by piaskowyk** [comment]
Did you tested it on Android?

**Comment by Latropos** [comment]
@piaskowyk Yes, it works on Andorid 💪

File: src/ReanimatedModule.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { NativeEventEmitter, NativeModules } from 'react-native';
+import { nativeShouldBeMock } from './reanimated2/PlatformChecker';
+import { StyleProps } from './reanimated2';
+
+export const NODE_MAPPING = new Map();
```
Comment (piaskowyk) [inline src/ReanimatedModule.ts:5]: 1. We don't want to export this outside the file. This is an internal module logic.
2. Can we rename `NODE_MAPPING` to something more descriptive?

File: src/ReanimatedModule.ts
Comment (piaskowyk) [inline src/ReanimatedModule.ts:1]: Can we rename file to something more descriptive?

File: src/createAnimatedComponent.tsx [hunk 3]
```diff
@@ -409,6 +416,22 @@ export default function createAnimatedComponent(
...
+        ReanimatedEventEmitter.addListener('onReanimatedPropsChange', listener);
+      }
+    }
+
+    _detachPropUpdater() {
+      const viewTag = findNodeHandle(this);
+      NODE_MAPPING.delete(viewTag);
+      if (NODE_MAPPING.size === 0) {
+        ReanimatedEventEmitter.removeAllListeners('onReanimatedPropsChange');
+      }
```
Comment (piaskowyk) [inline src/createAnimatedComponent.tsx:432]: same as above

File: src/createAnimatedComponent.tsx [hunk 4]
```diff
@@ -409,6 +416,22 @@ export default function createAnimatedComponent(
       }
     }
 
+    _attachPropUpdater() {
+      const viewTag = findNodeHandle(this);
+      NODE_MAPPING.set(viewTag, this);
+      if (NODE_MAPPING.size === 1) {
+        ReanimatedEventEmitter.addListener('onReanimatedPropsChange', listener);
+      }
```
Comment (piaskowyk) [inline src/createAnimatedComponent.tsx:424]: This logic should be in the file which you added.

File: src/ReanimatedModule.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+let reanimatedModule: typeof ReanimatedModuleMock;
+if (nativeShouldBeMock()) {
+  reanimatedModule = ReanimatedModuleMock;
+} else {
+  const { ReanimatedModule } = NativeModules;
+  reanimatedModule = ReanimatedModule;
+}
+
+export const ReanimatedEventEmitter = new NativeEventEmitter(reanimatedModule);
```
Comment (piaskowyk) [inline src/ReanimatedModule.ts:34]: we don't want to export `ReanimatedEventEmitter` outside the file. We want to export object which interface:
+ addOnJSPropsChangeListener(<component>)
+ removeOnJSPropsChangeListener(<viewtag>)

File: src/JSPropUpdater.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    TAG_TO_COMPONENT_MAPPING.set(viewTag, animatedComponent);
+    if (TAG_TO_COMPONENT_MAPPING.size === 1) {
+      this.ReanimatedEventEmitter.addListener(
+        'onReanimatedPropsChange',
+        listener
+      );
+    }
+  }
+
+  _detachPropUpdater(animatedComponent: React.Component<unknown, unknown>) {
```
Comment (piaskowyk) [inline src/JSPropUpdater.ts:55]:
```suggestion
  public detachPropUpdater(animatedComponent: React.Component<unknown, unknown>) {
```

**Review by piaskowyk — CHANGES_REQUESTED** [review]

File: src/JSPropUpdater.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    },
+  };
+
+  private static _listener(data: ListenerData) {
+    const component = JSPropUpdater._tagToComponentMapping.get(data.viewTag);
+    component && component._updateFromNative(data.props);
+  }
+
+  constructor() {
+    let _reanimatedModule: typeof JSPropUpdater._reanimatedModuleMock;
```
Comment (piaskowyk) [inline src/JSPropUpdater.ts:32]:
```suggestion
    let reanimatedModule: typeof JSPropUpdater._reanimatedModuleMock;
```

**Review by piaskowyk — CHANGES_REQUESTED** [review]

**Review by piaskowyk — APPROVED** [review]

File: src/createAnimatedComponent.tsx [hunk 8]
```diff
@@ -286,6 +287,7 @@ export default function createAnimatedComponent(
     _inlinePropsMapperId: number | null = null;
     _inlineProps: StyleProps = {};
     _sharedElementTransition: SharedTransition | null = null;
+    _JSPropUpdater = new JSPropUpdater();
```
Comment (tomekzaw) [inline src/createAnimatedComponent.tsx:290]: Probably `_jsPropsUpdater` would be a better name here

File: src/JSPropUpdater.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    if (nativeShouldBeMock()) {
+      reanimatedModule = JSPropUpdater._reanimatedModuleMock;
+    } else {
+      reanimatedModule = NativeModules.ReanimatedModule;
+    }
+    this._reanimatedEventEmitter = new NativeEventEmitter(reanimatedModule);
+  }
+
+  public addOnJSPropsChangeListener(
+    animatedComponent: React.Component<unknown, unknown>
```
Comment (tomekzaw) [inline src/JSPropUpdater.ts:42]: Is there any TypeScript way to enforce that the component must be animated?

File: src/JSPropUpdater.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      // noop
+    },
+    async removeListeners(): Promise<void> {
+      // noop
+    },
+  };
+
+  private static _listener(data: ListenerData) {
+    const component = JSPropUpdater._tagToComponentMapping.get(data.viewTag);
+    component && component._updateFromNative(data.props);
```
Comment (tomekzaw) [inline src/JSPropUpdater.ts:28]: I know this snippet of code was copied but it may be finally worth to convert it into a proper conditional.

**Comment by Latropos** [comment]
Not all svg issues were solved, looks that fill color still doesn't change https://github.com/software-mansion/react-native-reanimated/issues/1909

**Comment by tomekzaw** [comment]
As far as I know, it required a custom prop adapter, see `createAnimatedPropAdapter`.

**Pull request merged**